### PR TITLE
fix: content-driven language-mode divergence must not block compare

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Fixes a regression on `main` (PR #812, `6fb8536`) that made `compare` raise `ProfileMismatchError` instead of producing a verdict whenever the old and new sides of an unpinned header dump legitimately auto-detect into different C/C++ language modes.

## Motivation

`main`'s `CI` (`integration-tests` on all three OS legs) and `Examples Validation` workflows have been red since PR #812 merged. Root cause: #812 added `dumper_toolchain._probe_default_language_standard`/`_resolve_standard_provenance`, which now records a real `ast_resolved_standard` for an unpinned (`--lang` not given) header dump instead of always `None`. When old/new headers auto-detect into genuinely *different* C/C++ language modes purely from content — e.g. removing an `extern "C"` wrapper (`examples/case66_language_linkage_changed`) or adding a C++-only destructor (`examples/case69_trivial_to_nontrivial`) — the new probe trips the comparability gate and blocks the comparison outright, even though that divergence is exactly the ABI-relevant edit the two example cases exist to prove abicheck still catches.

## Changes

- Add `comparability._language_standard_content_divergence_corroborated`, a new carve-out that waives a `language_standard` mismatch when: neither side pinned `--lang` explicitly, the resolved language *mode* (`c`/`c++`) genuinely differs between old and new (`AbiSnapshot.ast_toolchain["resolved_lang_mode"]`), each side's `language_standard` provenance confirms it was never explicitly pinned via `-std=`/`--std=`/`/std:` either (`AbiSnapshot.ast_toolchain["language_standard_explicit"]`, newly stamped by `dumper_toolchain._stamp_ast_parser`), and the compiler identity (`compiler_family`, plus a driver-name-normalized `compiler_version`/`producer`/frontend `version` match) is independently confirmed unchanged.
- Deliberately **narrower** than "any unpinned `language_standard` divergence under a matching toolchain": a same-mode divergence caused only by a differing C++ *edition* (e.g. the `force_cpp20` requires-clause heuristic firing on only one side) is left blocked — that is exactly the toolchain-profile mismatch PR #812's probe exists to catch, and is pinned by the pre-existing `test_dumper_contract_wiring.py::test_cpp20_heuristic_forced_standard_flows_into_profile_fingerprint` regression test (which this change keeps passing).
- Real-toolchain fix found during review: under **castxml** specifically, a mode switch changes which host-compiler binary gets resolved (`gcc` for C mode, `g++` for C++), so their raw `--version` banners differ only in that leading driver-name word even for the identical toolchain installation. `_compiler_version_sans_driver_name` normalizes that away (a no-op for clang, and still correctly rejects a genuine cross-version GCC skew).
- New unit tests in `tests/test_comparability_gate_probe_upgrade.py` covering the waived case and every guard rail (differing compiler family/version/producer/frontend-version, explicit `--lang` pinned, explicit `-std=` given without `--lang`, the exact literal-collision case, missing provenance, same-mode-different-edition, and the castxml gcc/g++ driver-name normalization itself).
- Changelog fragment via `scriv create`.

## Bug-fix test contract

- Bug class: comparability-gate false positive — a content-driven language-*mode* divergence misclassified as a toolchain/extraction-environment mismatch, blocking a real, valid comparison.
- Publicly observable failure: `abicheck compare`/`scan` raises `ProfileMismatchError` (`NOT_COMPARABLE`) instead of producing a verdict for two unpinned header dumps whose headers legitimately differ in resolved C/C++ language mode — reproduced end-to-end against real compiled libraries for `examples/case66_language_linkage_changed` and `examples/case69_trivial_to_nontrivial`, matching the real CI failure signature on `main`.
- Regression test fails on base: yes — `test_gate_waives_content_driven_language_mode_divergence_when_toolchain_matches` raises `ProfileMismatchError` against the pre-fix code instead of passing; the two example cases raise the identical error end-to-end pre-fix.
- Negative control: `test_gate_content_driven_language_mode_divergence_still_raises_when_compiler_family_differs`, `_not_waived_when_lang_pinned`, `_not_waived_when_provenance_missing`, `test_gate_content_driven_divergence_still_raises_for_explicit_std_without_lang`/`_for_the_exact_literal_collision_pair`, `test_gate_content_driven_still_raises_when_producer_differs`/`_when_frontend_version_differs`/`_when_gnu_driver_version_number_genuinely_differs`, and `test_gate_same_mode_differing_edition_still_raises_even_with_compiler_unchanged` all confirm the new carve-out does not overreach.
- Public-surface test: exercised through `check_contracts_comparable`/`compare()`'s public entry point, and end-to-end via the real `abicheck dump`/`compare` CLI against real `g++`/clang-compiled `.so`s for both example cases.
- Real-dependency test: verified end-to-end against a **real castxml binary and real gcc/g++ 13.3.0** (matching the actual CI castxml lane's resolution mechanism, not a hand-constructed `AbiSnapshot`) reproducing `examples/case66_language_linkage_changed`: real `abicheck dump` + `compare` against real compiled `.so`s went from raising `ProfileMismatchError` (`differing fields: compiler_version`) to a real `BREAKING` verdict. This is what caught the driver-name-banner-skew bug in the first place — a synthetic `AbiSnapshot`-only test would not have caught it, since the bug was specifically in how castxml's *real* host-compiler resolution reports its version for `gcc` vs. `g++`.
- Axes covered: mode-switch waived only when toolchain matches; still blocked when compiler family/producer/frontend-version/normalized-host-version differ; still blocked when `--lang` was pinned explicitly on either side or an explicit `-std=` was given without `--lang`; still blocked when the mode is the same but the resolved *edition* differs (the `force_cpp20` case PR #812 added the probe to catch); verified on both the direct-clang backend and the real castxml backend (gcc/g++).
- General invariant: a content-driven `c`<->`c++` language-*mode* divergence under an identical, corroborated toolchain must never block comparability; a toolchain-driven divergence, an explicitly-pinned-language divergence, or a same-mode edition divergence must still block it.
- Known unsupported cases: none new — pre-existing toolchain-probing gaps already documented in `AGENTS.md`'s "Known gaps" section are unaffected.

## Checklist

- [x] Tests added / updated for new behaviour
- [x] Changelog fragment added (`scriv create`)
- [ ] Docs updated if user-visible behaviour changed (no user-visible behavior/doc change — internal comparability-gate correctness fix)
- [x] `pre-commit`-equivalent checks (`ruff check`, `ruff format --check` on touched files, `mypy abicheck/`) pass locally
- [x] No new `mypy` or `ruff` errors introduced

Verified against the full fast unit suite (29,059+ passed across three iterations of this fix) — only the pre-existing, unrelated `test_ai_readiness.py`/`test_cli_contract.py` failures remain, confirmed present identically on unmodified `main`. Also verified end-to-end against a real castxml binary + real gcc/g++ toolchain reproducing the exact CI failure signature.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01Xzg29XEzvhonN9iZC8cMmn)_

## Summary by CodeRabbit

- **Bug Fixes**
  - ABI comparisons now accept verified, content-driven C/C++ mode differences when compiler toolchains match.
  - Explicit settings, mismatched toolchains, insufficient provenance, and unverified compiler wrappers remain correctly flagged.
  - Concurrent comparison runs no longer overwrite shared report files.
  - Generated task checks are now consistent across platforms and line-ending formats.

- **Documentation**
  - Updated compatibility guidance and changelog information.

- **Tests**
  - Added coverage for language-mode provenance, compiler evidence, and comparison edge cases.

### CODE REVIEW

File: abicheck/comparability.py [hunk 1]
```diff
@@ -902,6 +902,125 @@ def _language_standard_probe_upgrade_corroborated(
...
+    ``compiler_sha256`` when both sides carry one), mirroring the sibling
+    carve-out's own toolchain-identity check.
+    """
+    old_std = old_fields.get("language_standard", "")
+    new_std = new_fields.get("language_standard", "")
+    if not old_std or not new_std or old_std == new_std:
+        return False
+    if _language_standard_is_lang_pinned(old_std) or _language_standard_is_lang_pinned(
+        new_std
+    ):
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/comparability.py]: Good catch — confirmed and fixed in 4c323e2. `_language_standard_is_lang_pinned` only recognizes an explicit lang *tag*, so an explicit `-std=gnu11`/`-std=gnu++17` given without `--lang` slipped through undetected, and `-std=gnu11` collides exactly with the bare literal the unpinned forced-default path also produces.

Added `_language_standard_is_known_auto_resolved_form`: a bare (non-`"probed:"`-prefixed) value is now only trusted when it exactly matches the one literal the *unpinned* path can actually produce for its own resolved mode (`gnu11` for `c`, `gnu++20` for `c++`); any other bare literal — which can only have come from an explicit pin — now correctly falls through to raising `ProfileMismatchError`. Two new regression tests pin both the general case and the exact literal-collision shape you flagged (`tests/test_comparability_gate_probe_upgrade.py::test_gate_content_driven_divergence_still_raises_for_explicit_std_without_lang` / `..._when_bare_forced_literal_is_actually_explicit`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/dumper_toolchain.py [hunk 2]
```diff
@@ -298,6 +316,9 @@ def _stamp_ast_parser(
         metadata["abi_dialect"] = dialect
     if resolved_force_cpp is not None:
         metadata["resolved_lang_mode"] = "c++" if resolved_force_cpp else "c"
+    metadata["language_standard_explicit"] = (
+        "1" if has_explicit_std(gcc_options, gcc_option_tokens) else "0"
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/dumper_toolchain.py:320]: Confirmed real (verified against a real compiler: `g++ -ansi` does report `__cplusplus=199711L`, not the default `201703L`) — but this is the same gap already investigated and explicitly documented as a deliberate "known gap, not fixed here" decision in `has_explicit_std`'s own docstring (`abicheck/_compiler_options.py:264-281`), from an earlier round of this same PR. The docstring gives the reason a scoped fix here would be wrong: `-ansi`'s resolved standard depends on the caller's language mode, which this function's two-argument signature doesn't currently receive, and every one of its several call sites across `dumper_ast_config.py`/`dumper_toolchain.py`/`cli_helpers_compare.py`/`service_compare_evidence.py`/`service_scan.py` would need re-verifying against a widened signature — a genuine, cross-cutting change, not a follow-up to one caller under continued review pressure (AGENTS.md's own "known gaps over risky reactive patches" convention). Tracked there rather than fixed reactively here.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/dumper_toolchain.py:320]: Confirmed real, and already investigated and documented as a deliberate "known gap, not fixed here" decision in `has_explicit_std`'s own docstring (`abicheck/_compiler_options.py:283-306`), from an earlier round of this same PR. Expanding a response file safely needs the same trust-boundary care this codebase already applies to `@file` expansion elsewhere (`depfile_args_from_argv`'s `trusted_root` parameter, per its own "Known gaps" entry in AGENTS.md) — resolving *this* file relative to *which* directory, and refusing to read outside it, is a real design question this function's filesystem-free signature has no way to answer today. A correct fix needs that trust boundary threaded into the signature and re-verified at every call site, not a same-PR patch under continued review pressure. Tracked there rather than fixed reactively here.

---
_Generated by [Claude Code](https://claude.ai/code)_
